File: robocode-tankroyale-sample-bots/java/Corners/src/main/java/dev/robocode/tankroyale/sample/bots/Corners.java
package dev.robocode.tankroyale.sample.bots;

import dev.robocode.tankroyale.botapi.Bot;
import dev.robocode.tankroyale.botapi.BotInfo;
import dev.robocode.tankroyale.botapi.events.BotDeathEvent;
import dev.robocode.tankroyale.botapi.events.ScannedBotEvent;

import java.io.IOException;

public class Corners extends Bot {

  int enemies; // Number of enemy robots in the game
  int corner = 0; // Which corner we are currently using
  volatile boolean stopWhenSeeRobot = false; // See goCorner()

  /** Main method starts our bot */
  public static void main(String[] args) throws IOException {
    new Corners().start();
  }

  /** Constructor, which loads the bot settings file */
  protected Corners() throws IOException {
    super(BotInfo.fromFile("corners.properties"));
  }

  /** This method runs our bot program, where each command is executed one at a time in a loop. */
  @Override
  public void run() {
    // Set colors
    setBodyColor("#F00"); // red
    setGunColor("#000"); // black
    setRadarColor("#FF0"); // yellow
    setBulletColor("#0F0"); // green
    setScanColor("#0F0"); // green

    // Save # of other bots
    enemies = getEnemyCount();

    // Move to a corner
    goCorner();

    // Initialize gun turn speed to 3
    int gunIncrement = 3;

    // Spin gun back and forth
    while (isRunning()) {
      for (int i = 0; i < 30; i++) {
        turnGunRight(gunIncrement);
      }
      gunIncrement *= -1;
    }
  }

  /** A very inefficient way to get to a corner. Can you do better? */
  private void goCorner() {
    // We don't want to stop when we're just turning...
    stopWhenSeeRobot = false;
    // turn to face the wall to the "right" of our desired corner.
    turnLeft(normalizeRelativeDegrees(corner - getDirection()));
    // Ok, now we don't want to crash into any robot in our way...
    stopWhenSeeRobot = true;
    // Move to that wall
    forward(5000);
    // Turn to face the corner
    turnRight(90);
    // Move to the corner
    forward(5000);
    // Turn gun to starting point
    turnGunRight(90);
  }

  /** We saw another bot. Stop and fire! */
  @Override
  public void onScannedBot(ScannedBotEvent e) {
    double distance = distanceTo(e.getX(), e.getY());

    // Should we stop, or just fire?
    if (stopWhenSeeRobot) {
      // Stop movement
      stop();
      // Call our custom firing method
      smartFire(distance);
      // Resume movement
      resume();
    } else {
      smartFire(distance);
    }
  }

  /**
   * Custom fire method that determines firepower based on distance.
   *
   * @param distance the distance to the robot to fire at
   */
  private void smartFire(double distance) {
    if (distance > 200 || getEnergy() < 15) {
      fire(1);
    } else if (distance > 50) {
      fire(2);
    } else {
      fire(3);
    }
  }

  /** We died. Figure out if we need to switch to another corner. */
  @Override
  public void onDeath(BotDeathEvent e) {
    // Well, others should never be 0, but better safe than sorry.
    if (enemies == 0) {
      return;
    }

    // If 75% of the robots are still alive when we die, we'll switch corners.
    if ((enemies - getEnemyCount()) / (double) enemies < .75) {
      corner += 90;
      if (corner == 270) {
        corner = -90;
      }
      System.out.println("I died and did poorly... switching corner to " + corner);
    } else {
      System.out.println("I died but did well. I will still use corner " + corner);
    }
  }
}

File: sample-bots/nim/Corners/Corners.nim
## Corners
##
## A sample bot originally made for Robocode by Mathew Nelson.
## Ported to Nim for Robocode Tank Royale.
##
## This robot moves to a corner, then rotates its gun back and forth
## scanning for enemies. If it performs poorly in a round, it will
## try a different corner in the next round.

import asyncdispatch
import random
import robocode_tank_royale

type
  Corners = ref object of Bot
    enemies: int
    corner: int
    stopWhenSeeEnemy: bool

method run(self: Corners) {.async.} =
  ## Called when a new round is started -> initialize and do some movement.
  ## Set colors
  await self.setBodyColor(Color.RED)
  await self.setTurretColor(Color.BLACK)
  await self.setRadarColor(Color.YELLOW)
  await self.setBulletColor(Color.GREEN)
  await self.setScanColor(Color.GREEN)

  ## Save # of other bots
  self.enemies = self.getEnemyCount()

  ## Move to a corner
  await self.goCorner()

  ## Initialize gun turn speed to 3
  var gunIncrement = 3.0

  ## Spin gun back and forth
  while self.isRunning():
    for i in 0..<30:
      await self.turnGunLeft(gunIncrement)
    gunIncrement *= -1.0

method goCorner(self: Corners) {.async.} =
  ## A very inefficient way to get to a corner.
  ## Can you do better as an home exercise? :)
  ## We don't want to stop when we're just turning...
  self.stopWhenSeeEnemy = false
  ## Turn to face the wall towards our desired corner
  await self.turnLeft(self.calcBearing(float(self.corner)))
  ## Ok, now we don't want to crash into any bot in our way...
  self.stopWhenSeeEnemy = true
  ## Move to that wall
  await self.forward(5000.0)
  ## Turn to face the corner
  await self.turnLeft(90.0)
  ## Move to the corner
  await self.forward(5000.0)
  ## Turn gun to starting point
  await self.turnGunLeft(90.0)

method onScannedBot(self: Corners, event: ScannedBotEvent) {.async.} =
  ## We saw another bot -> stop and fire!
  var distance = self.distanceTo(event.x, event.y)

  ## Should we stop, or just fire?
  if self.stopWhenSeeEnemy:
    ## Stop movement
    await self.stop()
    ## Call our custom firing method
    await self.smartFire(distance)
    ## Rescan for another bot
    await self.rescan()
    ## This line will not be reached when scanning another bot.
    ## So we did not scan another bot -> resume movement
    await self.resume()
  else:
    await self.smartFire(distance)

method smartFire(self: Corners, distance: float) {.async.} =
  ## Custom fire method that determines firepower based on distance.
  ## distance: The distance to the bot to fire at.
  if distance > 200.0 or self.getEnergy() < 15.0:
    await self.fire(1.0)
  elif distance > 50.0:
    await self.fire(2.0)
  else:
    await self.fire(3.0)

method onDeath(self: Corners, event: DeathEvent) {.async.} =
  ## We died -> figure out if we need to switch to another corner
  ## Well, others should never be 0, but better safe than sorry.
  if self.enemies == 0:
    return

  ## If 75% of the bots are still alive when we die, we'll switch corners.
  if self.getEnemyCount() / self.enemies.float >= 0.75:
    self.corner += 90 ## Next corner
    self.corner = self.corner mod 360 ## Make sure the corner is within 0 - 359

    echo "I died and did poorly... switching corner to ", self.corner
  else:
    echo "I died but did well. I will still use corner ", self.corner

proc randomCorner(): int =
  ## Returns a random corner (0, 90, 180, 270)
  ## Random number is between 0-3
  90 * rand(3)

when isMainModule:
  var corner = randomCorner()
  let bot = Corners(corner: corner)
  waitFor bot.start()